### DISCUSSION TITLE
xDrip+ should not use "media" for what are notifications.

### DISCUSSION THREAD

**Discussion by jbowler** [NightscoutFoundation/xDrip#2809] [Ideas]
See, for example:

https://navid200.github.io/xDrip/docs/AAPS_ProfileImportNotification.html

The "notification" isn't a notification; it's a media sound.  I know this because after suffering this interminable noise for a couple of weeks I determined that turning the _media_ volume to zero stops it while none of the other volume controls (notification, ring, alarmclock) on my 'phone had any effect.

Android apps which play a sound as some kind of notification/alarm have been around for years.  While I was trying to determine what was going on I did extensive web searches and both apps and, indeed, misplaced system sounds (e.g. battery notifications) were all fingered since at least 2014.  Finding out what app is doing this is almost impossible, hence my facebook post here (you may need to be an AndroidAPS FB member to see this):

https://www.facebook.com/groups/AndroidAPSUsers/permalink/3556203831267675

Yes; I really was on the point of returning my new, expensive, 'phone.  I really had deinstalled or disabled all the apps (including all of Google and including a lot of the semi-hidden apps).  It's took me about a day and other than a suspicion that it might be one of the only two apps I had left (xDrip+ and AndroidAPS) I was no further in my search.  Of course it did only take 10 minutes after my post to the FB ground for someone there to point a finger at xDrip+

This would NOT happen if xDrip+ was using a `notification` for a **notification**, because then I wouldn't have cared; notifications just go "beep" with my settings and I can disable them on a per-app basis.  They don't make a ridiculous trumpeting noise that I can't control!

It also would, apparently, not happen if the option was **not** turned on by default.

This is a classic, long standing, Android app programming error.  I would love it if Android actually had a "media" permission so that I could stop apps making media noises, but so far as I could determine it doesn't, so that just makes it worse.  Check on the web the number of people who have encountered this problem (in other apps/system components) and go to the point of insanity.

It can also be fixed by having an overall "turn off sounds" option.  I did find the option to set the BG level alerts all to "silent", but most of us will long there first.  If someone uses that option they are probably looking for the inverse, as I was; set all _other_ noises to silent.

**Comment by Navid200** [comment]
What do you want, as an xDrip user, to happen that currently does not?
There may already be a solution.

For example, do you want to hear no sound for your alerts?
Or what?

**Reply by jbowler** [reply]
I want notifications to be notifications.  I can suggest other fixes for this bug but the fundamental bug is that you are playing media for a notification.

**Reply by Navid200** [reply]
I want to know what functionality is missing.  I'm not asking about the meaning of notification versus media.
I believe you are an xDrip user.  Can you tell me what you need xDrip to do functionally that it doesn't?

It is quite possible that other users experience the same problem you are trying to describe.  But, if you don't describe it from the user perspective, we will never get anywhere.

You say you want notifications to be notifications.  That means nothing to me.  I am sorry.  
I'm happy to stay out of this thread.

**Reply by jbowler** [reply]
I "want" xDrip+ NOT to play media.  I.e. not to use the "media" channel in Android.  You have a choice of four (this is Android 13, but the distinction between media and notifications has existed pretty much for ever):

1) "Media"
2) "Ring"
3) "Notification"
4) "Alarm"

(3) and (4) are clearly in the province of xDrip+, "Ring" and "Media" are not.  Nothing in xDrip+ should be sending stuff down those channels.  As your own web page says, the problem is that a "Notification" is being handled as a "sound", "media".  As I explained in some detail Android doesn't have a way (that I can see) of detecting which app is generating "media", so when an app like xDrip+ starts playing random sounds, regardless of the programmer logic behind it, it is impossible for a user, even a very experienced SDE, to work out which app is doing it, let alone how to stop it.

**Reply by Navid200** [reply]
The alerts are notifications.   I like it that I can snooze the alert by swiping the associated notification.  I don't want that to change.
And I want the alert to wake me up.  So, I want it to make a sound.  
I don't want the alert volume to be controlled by the phone volume.  That's what volume profiles are for.

There are some things that need to improve.  
But, we cannot do that if we cannot communicate.

**Reply by Navid200** [reply]
As soon as I posted, I saw your post.
There should be no problem with snoozing an alert.  Are you saying you cannot?

**Reply by Navid200** [reply]
I think if you enable force speaker, it will use the Alarm channel.

**Reply by Navid200** [reply]
You can also use audio focus to reduce the volume of your media player when an xDrip alert goes off.

**Reply by jbowler** [reply]
> I like it that I can snooze the alert by swiping the associated notification. I don't want that to change.

It is **_NOT_** implemented as a notification, or (EDIT) per you subsequent replies as an alert (alarm).  It's an annoying noise.  There is no notification to swipe, there's no alert to silence; try it.

>I don't want the alert volume to be controlled by the phone volume.

Yes, but it **is**.  I think we are agreeing here; notifications are notifications and are controlled as notifications.  The "phone" volume (I assume you meant the Android "media" volume, not the "notification" volume) should not affect the volume of the notification, **but it does**; the opposite of your statement as quoted.

>And I want the alert to wake me up.

You just stated that you want to be woken up every time an AAPS profile update comes in.  My xDrip+ imports the AAPS profile every two to three minutes.  I'd post a screen shot but I've removed every app from my 'phone...  Yes.  I do know you don't want that and that you didn't _mean_ to say that, but you did say it because you quoted and spoke out of context.

To be clear, I am specifically talking about the AAPS profile update "sound" that is the web page link I gave.

**Reply by Navid200** [reply]
Looking at the title, xDrip+ should not use "media" for what are notifications, I misunderstood and thought this was about xDrip notifications, all of them.  All my questions and comments were based on that assumption.
I now know that this is only about a single notification.
You can ignore all my comments.

**Reply by jamorham** [reply]
@jbowler If you search for `Import Sounds` in xDrip settings or look in Inter App settings you can disable the profile import sound. When this was tested it only triggered on manual profile imports but I think in certain configurations or with newer versions of AAPS they can occur frequently as you describe. Previously it was discussed about making this setting default to off which I think would be a good idea based on your report. Thanks for bringing it to my attention.

**Reply by jbowler** [reply]
@jamorham I also have engineering mode switched on because I'm running xDrip+ on the "receiver" slot, however the person who first responded to my AAPS FB post had encountered the same problem.

I still maintain that this should be a regular notification, not an audio sound.  The user expectation for the two are very different and the UI guidelines are different.  The particular sound also seems to be full volume (i.e. the audio waveform is at the peak of the digital encoding).  This is rarely true of media and someone who is playing a quiet music at full volume is not going to be pleased when the xDrip+ audio kicks in, cuts out the music (on Android 12+) and deafens them.

**Comment by digiexchris** [comment]
Maybe I can help explain:

Rather than the app playing a sound as configured through xdrip's alert settings, xdrip should create a notification group, and create notification channels in those groups instead. Sounds or other types of audivble/visible alerting options then can be controlled via normal android tooling including do not disturb schedules and things like that built into android.

https://developer.android.com/develop/ui/views/notifications/channels

They may not have decided to go this route as the app currently supports api versions below 26 (Android 8).

**Reply by jbowler** [reply]
> They may not have decided to go this route as the app currently supports api versions below 26

xDrip+ currently (nightly build) has a number of notification channels.  I currently have "bgAlertChannel", "bgMissedAlertChannel", "nightscoutUploaderChannel" with active notifications but also "Uncategorized" which will presumably become disabled if xDrip+ moves to Android 8/version 26.

In this case a simple "AAPSChannel" would fit the bill, but apparently some work in this area will be required for Android 8.  I'm running Android 13.

**Reply by tomascalta** [reply]
Bro I have like 20 xdrip notification channels like 4 from each category for example "bgMissedAlertChannel", "bgMissedAlertChannel2", "bgMissedAlertChannel3" and so on and I have no idea whats going on. Dont you know by any chance dude? :D Im like loosing my mind here because of it

**Reply by jbowler** [reply]
It seems to be media audio so it bypasses the "notification" sounds.  If you turn media audio to 0 in the volume controls it should go away.  You can wind it back up to something sensible about 12 hours after activating a new Dexcom sensor (G6 or G7).  The Trumpet sound seems to be at max volume so you might want to be careful if wearing headphones during that time...

What I'm talking about only happens _once_ after inserting a new sensor.  It did seem to go away for a while but it came back a few sensors ago for me.  It's unpredictable (by me) when it will come out so maybe something random is going on or maybe it isn't media and is one of the other two channels - "ringtone" and "alarm", excluding "notification".

In recent Androids an up or down on the volume control presents a vertical bar with an ellipsis below it.  If you click on the ellipsis you get direct control over the four separate audio channels.  If you zap them all to 0 nothing should ever come out.  Leaving "ringtone" up should be ok.  I always keep "notifications" at 0 and "alarm" at a low level but "media" cannot be dropped unless you don't use the phone for phone calls (I don't) _and_ don't use it to listen to music (etc) - I _do_ do that.

**Reply by Navid200** [reply]
Glucose level alerts use volume profile.  They completely ignore the phone volume.
All the other alerts, like forecast low or persistent high alert, follow ring and notification volume.  

xDrip lets you use notification channels or not.  There is a setting you can use to control that.  If you enable notification channels, you can customize many of the notifications even the glucose level alerts.

There is no notification after starting a G7.  You can enable a notification for sensor expiry.  But, that would go off at the end of a sensor run not after starting one.  

There is not a single alert (notification) in xDrip that cannot be controlled.  Neither is there a single one that has a random behavior.
